base/reinterpretarray.jl

```diff
@@ -0,0 +1,127 @@
+            """))
+        end
+        isbits(T) || throwbits(S, T, T)
+        isbits(S) || throwbits(S, T, S)
+        (N != 0 || sizeof(T) == sizeof(S)) || throwsize0(S, T)
+        if N != 0 && sizeof(S) != sizeof(T)
+            dim = size(a)[1]
+            rem(dim*sizeof(S),sizeof(T)) == 0 || thrownonint(S, T, dim)
+        end
+        new{T, N, S, A}(a)
+    end
+end
+
+parent(a::ReinterpretArray) = a.parent
+
+eltype(a::ReinterpretArray{T}) where {T} = T
+function size(a::ReinterpretArray{T,N,S} where {N}) where {T,S}
+    psize = size(a.parent)
+    size1 = div(psize[1]*sizeof(S), sizeof(T))
+    tuple(size1, tail(psize)...)
+end
+
+unsafe_convert(::Type{Ptr{T}}, a::ReinterpretArray{T,N,S} where N) where {T,S} = Ptr{T}(unsafe_convert(Ptr{S},a.parent))
+
+@inline @propagate_inbounds getindex(a::ReinterpretArray{T,0}) where {T} = reinterpret(T, a.parent[])
+@inline @propagate_inbounds getindex(a::ReinterpretArray) = a[1]
+
+@inline @propagate_inbounds function getindex(a::ReinterpretArray{T,N,S}, inds::Vararg{Int, N}) where {T,N,S}
+    if sizeof(T) == sizeof(S)
+        return reinterpret(T, a.parent[inds...])
+    else
+        ind_start, sidx = divrem((inds[1]-1)*sizeof(T), sizeof(S))
+        t = Ref{T}()
+        s = Ref{S}()
+        @gc_preserve t s begin
+            tptr = Ptr{UInt8}(unsafe_convert(Ref{T}, t))
+            sptr = Ptr{UInt8}(unsafe_convert(Ref{S}, s))
+            i = 1
+            nbytes_copied = 0
+            # This is a bit complicated to deal with partial elements
+            # at both the start and the end. LLVM will fold as appropriate,
+            # once it knows the data layout
+            while nbytes_copied < sizeof(T)
+                s[] = a.parent[ind_start + i, tail(inds)...]
+                while nbytes_copied < sizeof(T) && sidx < sizeof(S)
+                    unsafe_store!(tptr, unsafe_load(sptr, sidx + 1), nbytes_copied + 1)
+                    sidx += 1
+                    nbytes_copied += 1
+                end
+                sidx = 0
+                i += 1
+            end
+        end
+        return t[]
+    end
+end
+
+@inline @propagate_inbounds setindex!(a::ReinterpretArray{T,0,S} where T, v) where {S} = (a.parent[] = reinterpret(S, v))
+@inline @propagate_inbounds setindex!(a::ReinterpretArray, v) = (a[1] = v)
+
+@inline @propagate_inbounds function setindex!(a::ReinterpretArray{T,N,S}, v, inds::Vararg{Int, N}) where {T,N,S}
+    v = convert(T, v)::T
+    if sizeof(T) == sizeof(S)
+        return setindex!(a.parent, reinterpret(S, v), inds...)
+    else
+        ind_start, sidx = divrem((inds[1]-1)*sizeof(T), sizeof(S))
+        t = Ref{T}(v)
+        s = Ref{S}()
+        @gc_preserve t s begin
+            tptr = Ptr{UInt8}(unsafe_convert(Ref{T}, t))
+            sptr = Ptr{UInt8}(unsafe_convert(Ref{S}, s))
+            nbytes_copied = 0
+            i = 1
+            @inline function copy_element()
```

Comment:
It seems like this'll severely penalize performance due to the impossibility of inferring the types of those closure variables

Reply:
Ok, I'll manually inline it.